src/Qir/Runtime/lib/QIR/QubitManagerRestrictedReuse.cpp

```diff
@@ -0,0 +1,468 @@
+    // TODO: Borrowing/returning are not yet supported.
+    if (requestedCapacity <= qubitCapacity)
+    {
+        return;
+    }
+
+    // Prepare new shared status array
+    int* newStatusArray = new int[requestedCapacity];
+    memcpy(newStatusArray, sharedQubitStatusArray, qubitCapacity * sizeof(newStatusArray[0]));
+    QubitListInSharedArray newFreeQubits(qubitCapacity, requestedCapacity - 1, newStatusArray);
+
+    // Set new data. All new qubits are added to the free qubits in the outermost area.
+    FreeQubitsCount += requestedCapacity - qubitCapacity;
+    delete[] sharedQubitStatusArray;
+    sharedQubitStatusArray = newStatusArray;
+    qubitCapacity = requestedCapacity;
+    freeQubitsInAreas[0].FreeQubitsReuseAllowed.MoveAllQubitsFrom(newFreeQubits, sharedQubitStatusArray);
+}
+
+bool CQubitManagerRestrictedReuse::IsDisabled(int id)
+{
+    return sharedQubitStatusArray[id] == DisabledMarker;
+}
+
+bool CQubitManagerRestrictedReuse::IsDisabled(Qubit qubit)
+{
+    return IsValid(qubit) && IsDisabled(QubitToId(qubit));
+}
+
+void CQubitManagerRestrictedReuse::Disable(Qubit qubit)
+{
+    FailIf(!IsValid(qubit), "Invalid qubit.");
+    int id = QubitToId(qubit);
+    // We can only disable explicitly allocated qubits that were not borrowed.
+    FailIf(!IsExplicitlyAllocated(id), "Cannot disable qubit that is not explicitly allocated.");
+    sharedQubitStatusArray[id] = DisabledMarker;
+    DisabledQubitsCount++;
+
+    AllocatedQubitsCount--;
+    FailIf(AllocatedQubitsCount < 0, "Incorrect count of allocated qubits.");
+}
+
+void CQubitManagerRestrictedReuse::Disable(Qubit* qubitsToDisable, int qubitCount)
+{
+    FailIf(qubitCount < 0, "Qubit count cannot be negative");
+    if (qubitsToDisable == nullptr || qubitCount == 0)
+    {
+        return;
+    }
+
+    for (int i = 0; i < qubitCount; i++)
+    {
+        Disable(qubitsToDisable[i]);
+    }
+}
+
+bool CQubitManagerRestrictedReuse::IsExplicitlyAllocated(long id)
+{
+    return sharedQubitStatusArray[id] == AllocatedMarker;
+}
+
+void CQubitManagerRestrictedReuse::ChangeStatusToAllocated(long id)
+{
+    FailIf(id >= qubitCapacity, "Internal Error: Cannot change status of an invalid qubit.");
+    sharedQubitStatusArray[id] = AllocatedMarker;
+    AllocatedQubitsCount++;
+    FreeQubitsCount--;
+}
+
+Qubit CQubitManagerRestrictedReuse::Allocate()
+{
+    long newQubitId = AllocateQubit();
+    if (newQubitId == NoneMarker && MayExtendCapacity)
+    {
+        EnsureCapacity(qubitCapacity * 2);
+        newQubitId = AllocateQubit();
+    }
+    FailIf(newQubitId == NoneMarker, "Not enough qubits.");
+    ChangeStatusToAllocated(newQubitId);
+    return CreateQubitObject(newQubitId);
+}
+
+Qubit* CQubitManagerRestrictedReuse::Allocate(long numToAllocate)
+{
+    FailIf(numToAllocate < 0, "Attempt to allocate negative number of qubits.");
+    if (numToAllocate == 0)
+    {
+        return new Qubit[0];
+    }
```

Comment:
It looks like a dusty/arguable corner of the C++ language when we allocate an array with `0` number of elements.
Consider documenting that case (there is a risk of undefined behavior when accessing the returned pointer, see "undefined behavior" [here](https://www.cplusplus.com/reference/new/operator%20new[]/)).

Reply:
This was a direct conversion from C#. And it wasn't a good idea to allocate array and return it anyway. I changed the code to accept the array as an input parameter instead. That way user may decide to use statically allocated (or stack allocated) array and avoid dynamic allocation altogether.